### PR TITLE
Lock flutter tool while updating artifacts

### LINKED ISSUE #2762: 'flutter' tool doesn't lock its cache dir when updating
If you run multiple copies of flutter simultaneously (e.g. because you're running it under `watch` in another window) and upgrade, all instances will try to download the new toolchain at once.

### PR DESCRIPTION
**Pull request by Hixie**
This prevents multiple simultaneous runs of the analyzer from stomping
over each other (e.g. multiple runs of 'update-packages'). Certain
long-lived commands (like analyze, run, logs) are exempted once they've
done enough work to be safe from most stomping action.

This still doesn't make us entirely safe from craziness, e.g. if you're
half way through an 'update-packages' run and you call 'git pull', who
knows what state you'll end up in. But there's only so much one can do.

Fixes https://github.com/flutter/flutter/issues/2762

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/analyze.dart [hunk 1]
```diff
@@ -287,6 +289,8 @@ class AnalyzeCommand extends FlutterCommand {
     server.onAnalyzing.listen((bool isAnalyzing) => _handleAnalysisStatus(server, isAnalyzing));
     server.onErrors.listen(_handleAnalysisErrors);
 
+    Cache.unlock();
```
Comment (abarth) [inline packages/flutter_tools/lib/src/commands/analyze.dart]: try / finally ?  What happens if there's an exception and we forget to unlock the file?

File: packages/flutter_tools/lib/src/commands/trace.dart [hunk 2]
```diff
@@ -57,6 +58,8 @@ class TraceCommand extends FlutterCommand {
       return 1;
     }
 
+    Cache.unlock();
```
Comment (abarth) [inline packages/flutter_tools/lib/src/commands/trace.dart]: What's with all these random unlocks?  What happens if we forget to have one?

**Comment by devoncarew** [comment]
@Hixie, did you consider doing very targeted locking around `Cache.updateAll()` and `flutter update-packages`? It seems like those would be the biggest offenders, and would mean that there'd be fewer places in the tool that had to be locking aware.

File: packages/flutter_tools/lib/src/commands/analyze.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/analyze.dart]: Made the comment in cache.dart longer.

File: packages/flutter_tools/lib/src/commands/trace.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/trace.dart]: Renamed to releaseLockEarly for clarity.

**Comment by Hixie** [comment]
@devon The problem is that anywhere that deals with the cache is potentially affected. For example, buildXcodeProject calls _addServicesToBundle calls _copyServiceFrameworks calls getServiceFromUrl calls cache.getThirdPartyFile. But buildXcodeProject also calls other functions that deal with the cache (e.g. looking at the engine build, which could change if another flutter process were to run at the same time). So now buildXcodeProject needs to know about the lock, so that all its work is in a coherent state. Same applies everywhere else.

That's why I went for the approach of just having nowhere know about the locking except the core runner. But doing that doesn't quite work either because of the long-running commands. So that's why those commands know to release the lock.

**Comment by devoncarew** [comment]
kk, got it. I think you're looking at making the whole tool safe, module parts of the long running commands, and I'm thinking of just making cache.updateAll() safe - just ensuring that two people don't try and update at the same time.

I'll take

**Comment by devoncarew** [comment]
Perhaps we could have the locking/unlocking be the responsibility of the abstract `FlutterCommand` class? Commands that don't op-in to locking throughout their lifetime could unlock just before their `runInProject` method is called?

**Comment by Hixie** [comment]
Right now the locking is done in `FlutterCommandRunner.runCommand`. I'm happy to move it to `FlutterCommand._run` if you think that would be better.

The only commands that need to unlock are the long-running ones. They presumably want to unlock just before they do their long-running thing, like calling the analyzer. That implies doing it in runInProject, rather than before. For example, analyze doesn't want to drop the lock until it's done reading all the `.packages` files that update-packages would otherwise stomp on.

**Comment by Hixie** [comment]
(I noticed that some tests weren't being run correctly when I was fixing the tests, which is what the extra bits in the patch are about.)

File: packages/flutter_tools/lib/src/cache.dart [hunk 3]
```diff
@@ -24,6 +24,54 @@ class Cache {
...
+  /// Lock the cache directory.
+  ///
+  /// This happens automatically on startup (see [FlutterCommandRunner.runCommand]).
+  ///
+  /// Normally the lock will be held until the process exits (this uses normal
+  /// POSIX flock semantics). Long-lived commands should release the lock by
+  /// calling [Cache.releaseLockEarly] once they are no longer touching the cache.
+  static Future<Null> lock() async {
+    if (!_lockEnabled)
+      return;
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/cache.dart]: `return null` here?

File: packages/flutter_tools/lib/src/runner/flutter_command_runner.dart [hunk 4]
```diff
@@ -173,10 +175,10 @@ class FlutterCommandRunner extends CommandRunner {
     if (globalResults['version']) {
       flutterUsage.sendCommand('version');
       printStatus(FlutterVersion.getVersion(Cache.flutterRoot).toString());
-      return new Future<int>.value(0);
+      return 0;
     }
 
-    return super.runCommand(globalResults);
+    return await super.runCommand(globalResults);
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/runner/flutter_command_runner.dart]: is `await` needed here?

**Comment by devoncarew** [comment]
ok, safety first. lgtm w/ a green travis

File: packages/flutter_tools/lib/src/runner/flutter_command_runner.dart (hunk 4 above)
Comment (Hixie) [inline packages/flutter_tools/lib/src/runner/flutter_command_runner.dart]: The super function returns a `Future<Foo>`, but we're in an async block so conceptually we're returning a `Foo`. Hence the await. I don't think it practice it makes any difference, I just like my types to match. :-)

File: packages/flutter_tools/lib/src/cache.dart (hunk 3 above)
Comment (Hixie) [inline packages/flutter_tools/lib/src/cache.dart]: fixed

**Comment by devoncarew** [comment]
lgtm

**Pull request merged**